### PR TITLE
add simple e2e for local deployment

### PR DESCRIPTION
**Pull request by pmorie**
Adds a simple e2e for local development, you can run it by doing `make test-e2e`. It will:

- Deploy the hub and spoke to the cluster pointed to by `$KUBECONFIG`
- Wait for the spoke to send CSR and approve it
- Wait for the spoke to create SpokeCluster and accept the spoke into the hub
- Wait for the spoke's status to show it is accepted.

Example output:

```shell
$ make test-e2e
kustomize build deploy/hub | kubectl apply -f -
namespace/open-cluster-management created
customresourcedefinition.apiextensions.k8s.io/spokeclusters.cluster.open-cluster-management.io created
serviceaccount/hub-sa created
clusterrole.rbac.authorization.k8s.io/system:open-cluster-management:hub created
clusterrolebinding.rbac.authorization.k8s.io/system:open-cluster-management:hub created
deployment.apps/hub-registration-controller created
cp /home/pmorie/go/src/github.com/open-cluster-management/registration/.kubeconfig dev-kubeconfig
kubectl config set clusters.kind-kind.server https://10.96.0.1 --kubeconfig dev-kubeconfig
Property "clusters.kind-kind.server" set.
kubectl create secret generic bootstrap-secret --from-file=kubeconfig=dev-kubeconfig -n open-cluster-management
secret/bootstrap-secret created
kustomize build deploy/spoke | kubectl apply -f -
namespace/open-cluster-management unchanged
serviceaccount/spoke-agent-sa created
clusterrolebinding.rbac.authorization.k8s.io/system:open-cluster-management:spoke created
secret/hub-kubeconfig-secret created
deployment.apps/spoke-agent created
go test ./test/e2e -v -- -ginko.v
=== RUN   TestE2E
Running Suite: E2E suite
========================
Random Seed: 1588980367
Will run 1 of 1 specs

• [SLOW TEST:42.040 seconds]
Loopback registration [development]
/home/pmorie/go/src/github.com/open-cluster-management/registration/test/e2e/loopback_test.go:21
  Should approve CSR and SpokeCluster
  /home/pmorie/go/src/github.com/open-cluster-management/registration/test/e2e/loopback_test.go:22
------------------------------

Ran 1 of 1 Specs in 42.041 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestE2E (42.04s)
PASS
ok      github.com/open-cluster-management/registration/test/e2e        42.050s
```

### CODE REVIEW

File: test/e2e/loopback_test.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+			spoke, err2 = spokeClusters.Get(context.TODO(), "local-development", metav1.GetOptions{})
+			return err2
+		})
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+		gomega.Expect(spoke.Spec.HubAcceptsClient).To(gomega.Equal(false))
+
+		ginkgo.By("Accepting the SpokeCluster")
+		spoke.Spec.HubAcceptsClient = true
+		err = retry.RetryOnConflict(spokeRetry, func() error {
+			var err2 error
```
Comment (deads2k) [inline test/e2e/loopback_test.go]: Need to get the latest inside or you perma-lool

File: test/e2e/e2e_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package e2e
+
+import (
+	"os"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
```
Comment (deads2k) [inline test/e2e/e2e_suite_test.go]: Kube usually leaves these as gomega and ginkgo

File: test/e2e/loopback_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		err = wait.Poll(1*time.Second, 45*time.Second, func() (bool, error) {
+			var err2 error
+			csrs, err2 = csrClient.List(context.TODO(), metav1.ListOptions{
+				LabelSelector: "open-cluster-management.io/cluster-name = local-development",
+			})
+			if err2 != nil {
+				return false, err2
+			}
+
+			if len(csrs.Items) >= 1 {
```
Comment (deads2k) [inline test/e2e/loopback_test.go:176]: If there are more than say five, we're hotlooping and have a bug

File: test/e2e/loopback_test.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+				return true, nil
+			}
+
+			return false, nil
+		})
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+
+		ginkgo.By("Approving the CSR")
+		csr := csrs.Items[0]
```
Comment (deads2k) [inline test/e2e/loopback_test.go]: I recommend approving them all in case there its looking for a particular one

File: test/e2e/loopback_test.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1beta1.CertificateSigningRequestCondition{
+			Type:    certificatesv1beta1.CertificateApproved,
+			Reason:  "Approved by E2E",
+			Message: "Approved as part of Loopback e2e",
+		})
+		_, err = csrClient.UpdateApproval(context.TODO(), &csr, metav1.UpdateOptions{})
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+
+		var (
+			spokeRetry = wait.Backoff{
```
Comment (deads2k) [inline test/e2e/loopback_test.go]: There out to be a decent default in the package

**Review by deads2k — CHANGES_REQUESTED** [review]
The retry one has to be fixed before merge. Great start.

File: test/e2e/loopback_test.go (hunk 3 above)
Comment (pmorie) [reply test/e2e/loopback_test.go:176]: I agree, do you think we should surface that here?

File: pkg/hub/csr/controller.go [hunk 6]
```diff
@@ -163,16 +165,3 @@ func isSpokeClusterClientCertRenewal(csr *certificatesv1beta1.CertificateSigning
 
 	return csr.Spec.Username == x509cr.Subject.CommonName
 }
-
-// Check whether a CSR is in terminal state
-func isCSRInTerminalState(status *certificatesv1beta1.CertificateSigningRequestStatus) bool {
```
Comment (deads2k) [inline pkg/hub/csr/controller.go:168]: nit, my personal preference is to expose from point of use and then directly use them in tests.  helper package grow weird deps which spider out

File: test/e2e/e2e_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+	clusterv1client "github.com/open-cluster-management/api/client/cluster/clientset/versioned"
+)
+
+func TestE2E(t *testing.T) {
+	gomega.RegisterFailHandler(ginkgo.Fail)
+	ginkgo.RunSpecs(t, "E2E suite")
+}
+
+var (
```
Comment (deads2k) [inline test/e2e/e2e_suite_test.go:25]: minor: we've found that creating a helper to produce a per instance client allow better tracking in traces as each can different agent strings

File: test/e2e/loopback_test.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+)
+
+var spokeNamespace string = ""
+
+var _ = ginkgo.Describe("Loopback registration [development]", func() {
+	ginkgo.It("Should register the hub as a spoke", func() {
+		var (
+			err    error
+			suffix = rand.String(6)
+			nsName = fmt.Sprintf("loopback-spoke-%v", suffix)
```
Comment (deads2k) [inline test/e2e/loopback_test.go:41]: use `GenerateName` instead

**Comment by deads2k** [comment]
I think some refinement in the factorization could help this be more re-useable, but it certainly proves the principle.

/lgtm
/approve

**Pull request merged**